File: sdk/php/devinsight.php
<?php
/**
 * DevInsight SDK for PHP
 *
 * Captures uncaught exceptions and fatal errors, reporting them to DevInsight.
 *
 * Usage:
 *   require_once 'devinsight.php'; // só isso
 *
 * Configuration (env vars or constants defined before require):
 *   DEVINSIGHT_API_KEY   — your API key
 *   DEVINSIGHT_BASE_URL  — base URL (default: https://devinsight-api.onrender.com)
 *   DEVINSIGHT_PROJECT   — project name (default: unknown)
 *   DEVINSIGHT_ENABLED=0 — disable SDK
 */

(static function (): void {
    $apiKey  = (string) (getenv('DEVINSIGHT_API_KEY')  ?: (defined('DEVINSIGHT_API_KEY')  ? DEVINSIGHT_API_KEY  : ''));
    $baseUrl = rtrim((string) (getenv('DEVINSIGHT_BASE_URL') ?: (defined('DEVINSIGHT_BASE_URL') ? DEVINSIGHT_BASE_URL : 'https://devinsight-api.onrender.com')), '/');
    $project = (string) (getenv('DEVINSIGHT_PROJECT')  ?: (defined('DEVINSIGHT_PROJECT')  ? DEVINSIGHT_PROJECT  : 'unknown'));
    $enabled = getenv('DEVINSIGHT_ENABLED') !== '0';

    if (!$enabled || $apiKey === '') {
        return;
    }

    $send = static function (string $message, string $type, string $stack) use ($apiKey, $baseUrl, $project): void {
        $body = (string) json_encode([
            'tipo'     => 'silent_backend_error',
            'mensagem' => $message,
            'contexto' => [
                'project_name'   => $project,
                'exception_type' => $type,
                'stack'          => $stack,
            ],
        ]);

        $url = $baseUrl . '/v1/diagnosticos';
        $headers = "Content-Type: application/json\r\nAuthorization: Bearer {$apiKey}";

        if (ini_get('allow_url_fopen')) {
            $ctx = stream_context_create([
                'http' => [
                    'method'        => 'POST',
                    'header'        => $headers,
                    'content'       => $body,
                    'timeout'       => 10,
                    'ignore_errors' => true,
                ],
            ]);
            @file_get_contents($url, false, $ctx);
        } elseif (function_exists('curl_init')) {
            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_POST           => true,
                CURLOPT_POSTFIELDS     => $body,
                CURLOPT_HTTPHEADER     => ['Content-Type: application/json', "Authorization: Bearer {$apiKey}"],
                CURLOPT_TIMEOUT        => 10,
                CURLOPT_RETURNTRANSFER => true,
            ]);
            curl_exec($ch);
            curl_close($ch);
        }
    };

    // Uncaught exceptions
    set_exception_handler(static function (\Throwable $e) use ($send): void {
        $send($e->getMessage(), \get_class($e), $e->getTraceAsString());
    });

    // Fatal errors (E_ERROR, E_PARSE, E_CORE_ERROR, E_COMPILE_ERROR)
    $fatalTypes = [E_ERROR, E_PARSE, E_CORE_ERROR, E_COMPILE_ERROR];
    register_shutdown_function(static function () use ($send, $fatalTypes): void {
        $error = error_get_last();
        if ($error !== null && \in_array($error['type'], $fatalTypes, true)) {
            $send(
                $error['message'],
                'FatalError',
                $error['file'] . ':' . $error['line']
            );
        }
    });
})();

File: sdk/ruby/devinsight.rb
# frozen_string_literal: true

# DevInsight SDK for Ruby
#
# Captures unhandled exceptions and reports them to DevInsight.
#
# Usage:
#   require_relative 'devinsight'  # só isso
#
# Configuration (env vars):
#   DEVINSIGHT_API_KEY   — your API key
#   DEVINSIGHT_BASE_URL  — base URL (default: https://devinsight-api.onrender.com)
#   DEVINSIGHT_PROJECT   — project name (default: unknown)
#   DEVINSIGHT_ENABLED=0 — disable SDK

require 'net/http'
require 'json'
require 'uri'

module Devinsight
  API_KEY  = ENV.fetch('DEVINSIGHT_API_KEY', '').freeze
  BASE_URL = ENV.fetch('DEVINSIGHT_BASE_URL', 'https://devinsight-api.onrender.com').chomp('/').freeze
  PROJECT  = ENV.fetch('DEVINSIGHT_PROJECT', 'unknown').freeze
  ENABLED  = ENV['DEVINSIGHT_ENABLED'] != '0'

  def self.send_diagnostic(message, exception_type, stack)
    return unless ENABLED && !API_KEY.empty?

    payload = JSON.generate(
      tipo: 'silent_backend_error',
      mensagem: message.to_s,
      contexto: {
        project_name:   PROJECT,
        exception_type: exception_type.to_s,
        stack:          stack.to_s
      }
    )

    uri  = URI("#{BASE_URL}/v1/diagnosticos")
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl     = uri.scheme == 'https'
    http.open_timeout = 10
    http.read_timeout = 10

    req = Net::HTTP::Post.new(uri.path.empty? ? '/' : uri.path)
    req['Content-Type']  = 'application/json'
    req['Authorization'] = "Bearer #{API_KEY}"
    req.body = payload

    http.request(req)
  rescue StandardError => e
    warn "[DevInsight] Falha ao enviar diagnóstico: #{e.message}"
  end

  # Register at_exit hook — $! holds the exception on unhandled crash in MRI Ruby 2.7+/3.x
  if ENABLED && !API_KEY.empty?
    at_exit do
      err = $!
      next if err.nil? || err.is_a?(SystemExit)

      Devinsight.send_diagnostic(
        err.message,
        err.class.name,
        err.backtrace&.join("\n") || ''
      )
    end
  end
end
